torch/utils/data/dataloader.py

```diff
@@ -99,6 +99,11 @@ class DataLoader(object):
         worker_init_fn (callable, optional): If not ``None``, this will be called on each
             worker subprocess with the worker id (an int in ``[0, num_workers - 1]``) as
             input, after seeding and before data loading. (default: ``None``)
+        auto_collation (bool, optional): usually auto collation mode is dynamically determined,
+            (i.e. if a ``batch_sampler`` is specified, ``auto_collation`` mode will be ``True``
```

Comment:
`auto_collation` is on if either `batch_size` or `batch_sampler` is specified.

Also, why is this needed?

Reply:
We are creating python bindings for the c++ chunkdataset API implementation already on pytorch code base (chunk.h). On that dataset implementation, a full batch is returned from the dataset (regardless of batch_sampler, sampler, batch_size, etc of python DataLoader class). 

Another way to go would be set batch_size=None on the Dataloader ctor, but that would be misleading for the users, as ChunkDataset receives batch_size as input and it is a positive number

Reply:
I will provide a test soon and this use will be clearer, but any feedback is welcome

Reply:
Thanks for the explanation. But I am still not convinced that adding this is better. It complicates the already pretty complicated auto-collation behavior. I.e., previously, auto collating means that the dataset is yielding individual samples, but now it doesn't!

Actually, for ChunkDataset, I don't think that you should use any of the DataLoader's auto-collation mechanism, including `collate_fn`. It should be clear that the dataset yields batched samples, and thus the dataloader should not do any collation. If you put `batch_size` in one place (dataset) and `collate_fn`+`auto_collation` in another (dataloader), you are essentially splitting the batching logic into two different places. It is against the philosophy of current dataloader design that the dataloader either handles all collation or not, so I think it would just be more confusing.

My suggestion is to just delegate all batching/collation to the `ChunkDataset`, i.e., putting `batch_size` and `collate_fn` to its ctor. And when one uses a `ChunkDataset`, they should understand what it does and set in dataloader ctor `batch_size=None`.

Reply:
Just as a context, this PR is roughly divided in two parts: 

- Adds new python bindings for existing C++ API (Chunkdataset was merged since Pytorch 1.1)
- Enhances Python Dataloader implementation to support the new ChunkDataset

This approach is interesting because C++ ChunkDataset API can support efficient data reading while python Dataloader allows an already familiar API for pytorch developers and flexible data transformation (such as torchvision.Transforms, numpy, etc) through `collate_fn`. Because some data transformation are very complicated to be done on C++, python bindings and integration with its DataLoader is a natural step for ChunkDataset. 

As you may have noticed, enhancing `auto-collation` is at the heart of the second part of the PR. I do agree DataLoader is already complex and, with that in mind, if `auto_collation` is not explicitly overridden by the user, the default behavior is exactly the same as before. That was done to prevent `DataLoader` from getting any more complicated for most users, but also to allow a finer granularity for advanced scenarios such as the new ChunkDataset API.

As you suggested, I am trying to disable auto-collation for `ChunkDataset(IterableDataset)`. By default, auto-collation is always on for `IterableDataset` as `batch_size=1` as per `DataLoader.__init__` (dataloader.py:116). As a consequence, `batch_sampler` will be assigned to `BatchSampler(sampler, batch_size, drop_last)` (dataloader.py:211). Because of that, `_auto_collation`property will return `True` (dataloader.py:243) and auto-collation will take place for `_IterableDatasetFetcher.fetch` (fetch.py:23). In the current implementation, the only way to prevent this from happening is to assign `batch_size=None`on DataLoader ctor, but that can be confusing for users of ChunkDataset API, as it does require a valid batch_size on its ctor. Given all the data flow for the current DataLoader implementation, my approach was to allow a manual toggle that disables auto-collation, preventing a confusing batch_size=None for Dataloader and batch_size>0 for ChunkDataset in the same script.

Regarding `collate_fn`, I have replied to it in another comment below (at dataloader.py:88)

@SsnL  Given this very long text explaining the intended data flow for `ChunkDataset` (sorry about that), do you still think that using two different values for `batch_size` (one `DataLoader ctor` and another for `ChunkDataset ctor`) is better than providing a setter for `Dataloader.auto_collation` that allows disabling it while keeping the default behavior intact?

Reply:
As per your suggestion, I have added a `collate_fn` to `ChunkDataset` class, which is called upon its `__next__` method.

Reply:
Removed auto_collation argument too